### PR TITLE
Add extension data support to Json serializer

### PR DESCRIPTION
**Pull request by steveharter**
Replaces PR https://github.com/dotnet/corefx/pull/37578 which had issues with a corrupt branch.

Adds the overflow or data extension support via the new [JsonExtensionDataAttribute](https://github.com/dotnet/corefx/issues/37558)

Other changes:
- Fixes https://github.com/dotnet/corefx/issues/37227 which allows using a Dictionary without a public setter.
- Refactors the "ignore" code and some tests added.
- Adds some additional validation around Dictionary (extension data uses a Dictionary).

Open issues:
- We do not apply serializer-only options to the reader\writer:
  - A property naming policy, such as a possible `ExtensionDataPropertyNaming` property. The existing naming policies for normal properties and dictionary keys.
  - IgnoreNullValues.
- This does not support Stream scenarios yet where the current buffer may run out of data. This will be addressed along with similar cases where this is encountered including the object->JsonElement converter and the extensibility model where the reader is exposed.

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                jsonPropertyInfo.SetValueAsObject(state.Current.ReturnValue, extensionData);
+            }
+
+            JsonElement jsonElement;
+            using (JsonDocument jsonDocument = JsonDocument.ParseValue(ref reader))
+            {
+                jsonElement = jsonDocument.RootElement.Clone();
+            }
+
+            string keyName = Encoding.UTF8.GetString(unescapedPropertyName);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs]: This overload doesn't exist on netstandard2.0. Hence the build failure on netfx:

> System\Text\Json\Serialization\JsonSerializer.Read.HandlePropertyName.cs(124,54): error CS1503: Argument 1: cannot convert from 'System.ReadOnlySpan<byte>' to 'byte[]'
System\Text\Json\Serialization\JsonSerializer.Write.HandleDictionary.cs(164,41): error CS1503: Argument 1: cannot convert from 'string' to 'System.ReadOnlySpan<char>'

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs [hunk 2]
```diff
@@ -156,5 +152,21 @@ public static partial class JsonSerializer
...
+
+        private static void WriteExtensionData(Utf8JsonWriter writer, ref WriteStackFrame frame)
+        {
+            DictionaryEntry entry = ((IDictionaryEnumerator)frame.Enumerator).Entry;
+            if (entry.Value is JsonElement element)
+            {
+                Debug.Assert(entry.Key is string);
+
+                string propertyName = (string)entry.Key;
+                element.WriteAsProperty(propertyName, writer);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: Similarly, I think we may need the explicit call to `AsSpan()` (since the span to string implicit converter is on core only)

We should just add the `WriteAsProperty` string-based overload to JsonElement to avoid this concern all together.

> System\Text\Json\Serialization\JsonSerializer.Write.HandleDictionary.cs(164,41): error CS1503: Argument 1: cannot convert from 'string' to 'System.ReadOnlySpan<char>'

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs [hunk 3]
```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
...
-                else
-                {
-                    NameUsedToCompareAsString = NameAsString;
-                    _nameUsedToCompare = _name;
+                    ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameNull(ParentClassType, this);
                 }
+            }
+            else
+            {
+                NameAsString = PropertyInfo.Name;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: And `PropertyInfo.Name` is guaranteed to not be null, correct?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs [hunk 4]
```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
...
+            }
+
+            // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
+            _name = Encoding.UTF8.GetBytes(NameAsString);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: From https://github.com/dotnet/corefx/pull/37578#discussion_r283076379

> Also, is `_name` a private property here rather than a field? Why? Maybe using PascalCasing or make it a field?

> `_name` is a private field.

`_name` has a getter/setter. It looks like its a property to me: https://github.com/dotnet/corefx/blob/33247077de38f44d68d18e5b3bd11fe341f28bd0/src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs#L28

If the intention is for it to be a field, I am fine with that, but we should remove the getter/setter.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs [hunk 5]
```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
...
+                NameUsedToCompareAsString = NameAsString;
+                _nameUsedToCompare = _name;
+            }
 
-                // Cache the escaped name.
+            // Cache the escaped name.
 #if true
-                // temporary behavior until the writer can accept escaped string.
-                _escapedName = _name;
+            // temporary behavior until the writer can accept escaped string.
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: Is this related to `JsonEncodedText`? We already added such overloads to the writer.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs [hunk 6]
```diff
@@ -94,100 +94,91 @@ public virtual void GetPolicies(JsonSerializerOptions options)
...
-                int valueIdx = JsonWriterHelper.NeedsEscaping(_name);
-                if (valueIdx == -1)
-                {
-                    _escapedName = _name;
-                }
-                else
-                {
-                    byte[] pooledName = null;
-                    int length = JsonWriterHelper.GetMaxEscapedLength(_name.Length, valueIdx);
+            int valueIdx = JsonWriterHelper.NeedsEscaping(_name);
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: In any case, I don't think the #else code block is adding much value, and we should get rid of it, for now.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs [hunk 7]
```diff
@@ -265,11 +257,25 @@ public void CopyRuntimeSettingsTo(JsonPropertyInfo other)
             other._escapedName = _escapedName;
         }
 
+        // Create a property that is either ignored at run-time. It uses typeof(int) in order to prevent
+        // issues with unsupported types and helps ensure we don't accidently (de)serialize it.
+        public static JsonPropertyInfo CreateIgnoredPropertyPlaceholder(PropertyInfo propertyInfo, JsonSerializerOptions options)
+        {
+            JsonPropertyInfo jsonPropertyInfo = new JsonPropertyInfoNotNullable<int, int, int>();
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: Would this end up creating a new `JsonPropertyInfoNotNullable` every time? Should this be a static field which is initialized once and re-used for all ignored properties?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs [hunk 8]
```diff
@@ -15,29 +14,17 @@ internal sealed class JsonPropertyInfoNotNullable<TClass, TDeclaredProperty, TRu
...
+            Debug.Assert(ShouldDeserialize);
+
             if (ElementClassInfo != null)
             {
                 // Forward the setter to the value-based JsonPropertyInfo.
                 JsonPropertyInfo propertyInfo = ElementClassInfo.GetPolicyProperty();
                 propertyInfo.ReadEnumerable(tokenType, options, ref state, ref reader);
             }
-            else if (ShouldDeserialize)
+            else
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs]: I still think we should avoid such deep nesting here:
https://github.com/dotnet/corefx/pull/37578#discussion_r283076814

File: src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs [hunk 9]
```diff
@@ -37,17 +36,45 @@ internal struct ReadStackFrame
         public bool Drain;
 
         public bool IsDictionary => JsonClassInfo.ClassType == ClassType.Dictionary;
+
+        public bool IsDictionaryProperty
+        {
+            get
+            {
+                if (JsonPropertyInfo != null && !JsonPropertyInfo.IsPropertyPolicy && JsonPropertyInfo.ClassType == ClassType.Dictionary)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs]: nit:
```C#
public bool IsDictionaryProperty
   => (JsonPropertyInfo != null &&
         !JsonPropertyInfo.IsPropertyPolicy &&
         JsonPropertyInfo.ClassType == ClassType.Dictionary);
```

OR with the getter syntax, just `return <condition>`.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: Did you start a separate discussion to add the public API?

For now, I'll just do the `AsSpan()`. When the overload is added we will need to check for all usages and clean up as appropriate.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs (hunk 1 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs]: Yep. Thanks

File: src/System.Text.Json/src/Resources/Strings.resx [hunk 10]
```diff
@@ -342,4 +342,19 @@
   <data name="DeserializeTypeNotSupported" xml:space="preserve">
     <value>Deserialization of type {0} is not supported.</value>
   </data>
+  <data name="SerializationDataExtensionPropertyInvalid" xml:space="preserve">
+    <value>The data extension property '{0}.{1}' does not match the required signature of IDictionary&lt;string, JsonElement&gt; or IDictionary&lt;string, object&gt;.</value>
+  </data>
+  <data name="SerializationDuplicateAttribute" xml:space="preserve">
+    <value>The attribute '{0}' cannot exist on more than one property.</value>
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/System.Text.Json/src/Resources/Strings.resx]: Is this better: `A class cannot have more than one property that has the attribute '{0}'.`

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 11]
```diff
@@ -308,37 +350,40 @@ private static ulong GetKey(ReadOnlySpan<byte> propertyName)
...
+            }
+
+            // Check for Dictionary<TKey, TValue> or IEnumerable<T>
+            if (propertyType.IsGenericType)
             {
-                elementType = propertyType.GetElementType();
-                if (elementType == null)
+                Type[] args = propertyType.GetGenericArguments();
+
+                if (GetClassType(propertyType) == ClassType.Dictionary &&
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Yes thanks.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 3 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: Yes. The beginning of the method checks for it.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 4 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: You are correct. It is a property. At one point it was a field. This data is on the "hot path" of the stack-based state so I was using fields which are faster than properties, so there is\was a mix of field access and some protected property setter for cases that aren't on the "hot path". However I find that doing this is causing more trouble in reviews than it is worth so I will make them all properties (if accessed outside the class) and measure perf impact at a later stage when we micro-optimize.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 5 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: Yes we need to convert this code to use that new feature (outside of this PR).

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 6 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: I'll do this when we convert to use that new feature.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 7 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: It needs to have its name set, so it can't be a global.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs (hunk 8 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs]: I commented on the link - the code is also trying to prevent pre-initialization of `value`.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 4 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: I have no concerns with it being fields (especially if you previously observed perf difference). I have just gotten used to properties being pascal cased, so having a "field-like" name threw me off.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs (hunk 7 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs]: Ah right. Good point.

File: src/System.Text.Json/src/Resources/Strings.resx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/System.Text.Json/src/Resources/Strings.resx]: > Is this better

Yes.

However, re-thinking about this, what happens when we add struct support? Is there a generic term that can encapsulate both .NET structs and classes? Type? Object?

Maybe type definition?
`A type definition cannot have more than one property that has the attribute '{0}'.`

**Comment by steveharter** [comment]
There is some feedback that will be addressed in a  future commit.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs]: > Did you start a separate discussion to add the public API?

Yes. Added the overload. See: https://github.com/dotnet/corefx/pull/37789
